controllers/tenant/rolebindings.go

```diff
@@ -160,3 +160,160 @@ func (r *Manager) syncAdditionalRoleBinding(ctx context.Context, tenant *capsule
+	var subject capsulev1beta2.ExtendedSubject
+
+	if owner.Kind == "ServiceAccount" {
+		splitName := strings.Split(owner.Name, ":")
+
+		subject = capsulev1beta2.ExtendedSubject{
+			Subject: rbacv1.Subject{
+				Kind:      owner.Kind.String(),
+				Name:      splitName[len(splitName)-1],
+				Namespace: splitName[len(splitName)-2],
+			},
+			// The owner should by default act as owner
+			ActAsOwner: true,
+		}
+	} else {
+		subject = capsulev1beta2.ExtendedSubject{
+			Subject: rbacv1.Subject{
+				APIGroup: rbacv1.GroupName,
+				Kind:     owner.Kind.String(),
+				Name:     owner.Name,
+			},
+			// The owner should by default act as owner
+			ActAsOwner: true,
+		}
+	}
+
+	return capsulev1beta2.PermissionSpec{
+		RoleBindings: clusterRoles,
+		Subjects: []capsulev1beta2.ExtendedSubject{
+			subject,
+		},
+	}
+}
+
+// Sync the dynamic Permissions specific cluster-roles and role bindings.
+func (r *Manager) syncPermissions(ctx context.Context, tenant *capsulev1beta2.Tenant) (err error) {
+
+	// hashing the RoleBinding name due to DNS RFC-1123 applied to Kubernetes labels
+	hashFn := func(binding capsulev1beta2.PermissionSpec) string {
+		h := fnv.New64a()
+
+		for _, cr := range binding.RoleBindings {
+			_, _ = h.Write([]byte(cr))
+		}
+
+		for _, sub := range binding.Subjects {
+			_, _ = h.Write([]byte(sub.Kind + sub.Name))
+		}
+
+		return fmt.Sprintf("%x", h.Sum64())
+	}
+	// getting requested Role Binding keys
+	keys := make([]string, 0, len(tenant.Spec.Owners))
+	// Generating for dynamic tenant owners cluster roles
+	for _, owner := range tenant.Spec.Owners {
+		cr := r.ownerClusterRoleBindingsToPermissions(owner, owner.ClusterRoles)
+
+		keys = append(keys, hashFn(cr))
+	}
+
+	// Generating hash of additional role bindings
+	for _, i := range tenant.Spec.Permissions {
+		keys = append(keys, hashFn(i))
+	}
+
+	group := new(errgroup.Group)
+
+	for _, ns := range tenant.Status.Namespaces {
+		namespace := ns
+
+		group.Go(func() error {
+			return r.syncPermissionsRoleBindings(ctx, tenant, namespace, keys, hashFn)
+		})
+	}
+
+	return group.Wait()
+}
+
+//nolint:nakedret
+func (r *Manager) syncPermissionsRoleBindings(ctx context.Context, tenant *capsulev1beta2.Tenant, ns string, keys []string, hashFn func(binding capsulev1beta2.PermissionSpec) string) (err error) {
+
+	var tenantLabel, roleBindingLabel string
+
+	if tenantLabel, err = utils.GetTypeLabel(&capsulev1beta2.Tenant{}); err != nil {
+		return
+	}
+
+	if roleBindingLabel, err = utils.GetTypeLabel(&rbacv1.RoleBinding{}); err != nil {
+		return
+	}
+
+	if err = r.pruningResources(ctx, ns, keys, &rbacv1.RoleBinding{}); err != nil {
+		return
+	}
+
+	var roleBindings []capsulev1beta2.PermissionSpec
+
+	for _, owner := range tenant.Spec.Owners {
+		roleBindings = append(roleBindings, r.ownerClusterRoleBindingsToPermissions(owner, owner.ClusterRoles))
+	}
+
+	for i, roleBinding := range roleBindings {
```

Comment:
Refactor reconcile per binding into dedicated function.
Use [goroutines](https://go.dev/tour/concurrency) to make execution parallel